rust/src/pgsql/parser.rs

```diff
@@ -0,0 +1,191 @@
+                assert_eq!(packet.is_valid(), true);
+            }
+            Err(nom::Err::Error((remainder, err))) => {
+                panic!("Result should not be an error: {:?}.", err);
+            }
+            Err(nom::Err::Incomplete(_)) => {
+                panic!("Result should not have been incomplete.");
+            }
+            _ => {
+                panic!("Unexpected behavior!");
+            }
+        }
+
+    }
+
+    #[test]
+    fn test_parse_pgsql_startup_packet_invalid_input() {
+        // A non-valid Startup message. Length field is too short
+        let invalid_length_field_buff:&[u8] = &[
+        /* Length (7) */ 0x00, 0x00, 0x00, 0x07,
+        /* Protocol Version */ 0x00, 0x03, 0x00];
+
+        let result = parse_pgsql_startup_packet(&invalid_length_field_buff);
+
+        match result {
+            Ok((remainder, packet)) => {
+                // this packet has length field too short so it is not valid
+                assert_eq!(packet.is_valid(), false);
+            }
+            Err(nom::Err::Error((remainder, err))) => {
+                panic!("Result should not be an error: {:?}.", err);
+            }
+            Err(nom::Err::Incomplete(needed)) => {
+                assert!(invalid_length_field_buff.len() < 
+                        (PGSQL_PROTO_SIZE + PGSQL_LENGTH_SIZE));
+            }
+            _ => {
+                panic!("Unexpected behavior!");
+            }
+        }
+
+        // A non-valid Startup message. Protocol version not allowed
+        let invalid_proto_version_buff:&[u8] = &[
+        /* Length (38) */       0x00, 0x00, 0x00, 0x26,
+        /* Protocol Version */  0x00, 0x12, 0x00, 0x00,
+        /* Data */  0x75, 0x73, 0x65, 0x72, 0x00, 0x69, 0x6e, 0x64, 0x65, 0x78,
+                    0x65, 0x72, 0x00, 0x64, 0x61, 0x74, 0x61, 0x62, 0x61, 0x73,
+                    0x65, 0x00, 0x69, 0x6e, 0x64, 0x65, 0x78, 0x65, 0x72, 0x00];
+
+        let result = parse_pgsql_startup_packet(&invalid_proto_version_buff);
+
+        match result {
+            Ok((remainder, packet)) => {
+                // this packet has a protocol major version not supported ( > 3)
+                assert_eq!(packet.is_valid(), false);
+
+                // length shouldn't be an issue here
+                assert_eq!(packet.length, 38);
+
+                // there should be nothing left
+                assert_eq!(remainder.len(), 0);
+            }
+            Err(nom::Err::Error((remainder, err))) => {
+                panic!("Result should not be an error: {:?}.", err);
+            }
+            Err(nom::Err::Incomplete(needed)) => {
+                panic!("Should be Ok, but needs {:?}", needed);
+            }
+            _ => {
+                panic!("Unexpected behavior!");
+            }
+        }
+
+        // A non-valid Startup message. Data length < length field indicates
+        let invalid_data_length_buff:&[u8] = &[
```

Comment:
can you also add a test like this with a length field that is < 8? I suspect this will underflow the parser and trigger a panic.

Reply:
The difference with the test at line 111 would be that there the nom parser exists with `Incomplete` before it hits that code.

Reply:
The part in lines 122 to 144 weas supposed to do that. But I wasn't sure how to handle the result, so I added an                 `assert!(invalid_length_field_buff.len() < (PGSQL_PROTO_SIZE + PGSQL_LENGTH_SIZE));` to the Incomplete arm of the match, for that's where my test goes...